tools/tx-scanner/cmd/scan.go

```diff
@@ -0,0 +1,153 @@
+	cmd.PersistentFlags().Int("port", 9090, "GRPC server port")
+	cmd.PersistentFlags().Int("batch-size", 100, "Batch size to query")
+	cmd.PersistentFlags().Int("bps-limit", 400, "Blocks per second limit")
+	cmd.PersistentFlags().Int64("start-height", 0, "Start height")
+	cmd.PersistentFlags().String("state-dir", "", "State file directory, the scanner will record the last scanned offset and scan results")
+	return cmd
+}
+
+func execute(cmd *cobra.Command, _ []string) {
+	endpoint, _ := cmd.Flags().GetString("endpoint")
+	port, _ := cmd.Flags().GetInt("port")
+	bpsLimit, _ := cmd.Flags().GetInt("bps-limit")
+	batchSize, _ := cmd.Flags().GetInt("batch-size")
+	stateDir, _ := cmd.Flags().GetString("state-dir")
+	startHeight, _ := cmd.Flags().GetInt64("start-height")
+	var badBlocks []int64
+	var currentState = state.State{}
+	if batchSize > bpsLimit {
+		batchSize = bpsLimit
+	}
+	if startHeight <= 0 {
+		startHeight = 1
+	}
+	if stateDir != "" {
+		scanState, _ := state.ReadState(stateDir)
+		if scanState.LastProcessedHeight > 0 {
+			fmt.Printf("Detected last processed height: %d\n", scanState.LastProcessedHeight)
+			currentState = scanState
+			startHeight = currentState.LastProcessedHeight
+			badBlocks = currentState.BlocksMissingTxs
+		}
+	}
+	fmt.Printf("Starting the scan from height: %d\n", startHeight)
+	client.InitializeGRPCClient(endpoint, port)
+	rateLimiter := rate.NewLimiter(rate.Limit(bpsLimit), bpsLimit)
+	latestHeight := getLatestBlockHeight()
+	var currBlockHeight = startHeight
+	for {
+		if currBlockHeight >= latestHeight {
+			time.Sleep(10 * time.Second)
+			latestHeight = getLatestBlockHeight()
+			continue
+		}
+		if !rateLimiter.AllowN(time.Now(), batchSize) {
+			time.Sleep(1 * time.Millisecond)
+			continue
+		}
+
+		wg := sync.WaitGroup{}
+
+		// Handle ALL the queries in a batch concurrently
+		var adjustedBatchSize = int(math.Min(float64(batchSize), float64(latestHeight-currBlockHeight)))
+		var mtx = sync.Mutex{}
+		var errors []error
+		for i := 0; i < adjustedBatchSize; i++ {
+			height := currBlockHeight + int64(i)
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				isBad, err := processBlock(height)
+				mtx.Lock()
+				defer mtx.Unlock()
+				if err != nil {
+					errors = append(errors, err)
+					return
+				}
+				if isBad {
+					badBlocks = append(badBlocks, height)
+				}
+			}()
+		}
+		// Wait for ALL queries in this batch to finish and then check any failures
+		wg.Wait()
+		if len(errors) > 0 {
+			fmt.Printf("Failed to process some blocks between heights %d and %d\n", currBlockHeight, currBlockHeight+int64(adjustedBatchSize))
+		} else {
+			// update the state
+			currBlockHeight += int64(adjustedBatchSize)
+			currentState.LastProcessedHeight = currBlockHeight
+			currentState.BlocksMissingTxs = badBlocks
+			if stateDir != "" {
+				err := state.WriteState(stateDir, currentState)
+				if err != nil {
+					fmt.Println(err.Error())
```

Comment:
should this return?

Reply:
Good questions, maybe it's better to panic if we failed to write to the state file